### PR TITLE
Change genesis preserving contract history

### LINKED ISSUE #878: Preserve contract history/ created date on genesis import
For history reasons with Tendermint starting at height 0 only, we had squashed the contract history on dump state import. This is not needed anymore as Tendermint evolved and leaves some smell with tests using imports.

Preserving the full history and state would also enable adding more data to queries as [requested feature](https://github.com/CosmWasm/cosmwasm/issues/1295).

### PR DESCRIPTION
**Pull request by GNaD13**
Close #878

### CODE REVIEW

File: x/wasm/keeper/genesis.go [hunk 1]
```diff
@@ -107,8 +107,6 @@ func ExportGenesis(ctx sdk.Context, keeper *Keeper) *types.GenesisState {
 			state = append(state, types.Model{Key: key, Value: value})
 			return false
 		})
-		// redact contract info
-		contract.Created = nil
```
Comment (alpe) [inline x/wasm/keeper/genesis.go:111]: 👍

File: app/sim_test.go [hunk 2]
```diff
@@ -211,22 +212,6 @@ func TestAppImportExport(t *testing.T) {
 	dropContractHistory(ctxA.KVStore(app.keys[wasm.StoreKey]), prefixes...)
 	dropContractHistory(ctxB.KVStore(newApp.keys[wasm.StoreKey]), prefixes...)
 
-	normalizeContractInfo := func(ctx sdk.Context, app *WasmApp) {
```
Comment (alpe) [inline app/sim_test.go:214]: 👍

**Review by alpe — COMMENTED** [review]
Very nice work! Thanks a lot for the changes 👍 

You addressed the `created` date proper. Unfortunately this was only half of the original story. The other part is about the `ContractHistory`. The history contains all the code changes for a contract. We don't export it but create a new start on [import](https://github.com/CosmWasm/wasmd/blob/main/x/wasm/keeper/keeper.go#L1077) . 
if you can add this to this PR, then #878  would be resolved. But also happy to merge this one and have a second to finish the issue. What you prefer.

**Comment by GNaD13** [comment]
> 

Got it, I will continue with the Contract History. Tysm :3

**Comment by GNaD13** [comment]
> Very nice work! Thanks a lot for the changes +1
> 
> You addressed the `created` date proper. Unfortunately this was only half of the original story. The other part is about the `ContractHistory`. The history contains all the code changes for a contract. We don't export it but create a new start on [import](https://github.com/CosmWasm/wasmd/blob/main/x/wasm/keeper/keeper.go#L1077) . if you can add this to this PR, then #878 would be resolved. But also happy to merge this one and have a second to finish the issue. What you prefer.

Got it, I will continue with the Contract History. Tysm :3

File: proto/cosmwasm/wasm/v1/genesis.proto [hunk 3]
```diff
@@ -53,6 +53,7 @@ message Contract {
   string contract_address = 1;
   ContractInfo contract_info = 2 [ (gogoproto.nullable) = false ];
   repeated Model contract_state = 3 [ (gogoproto.nullable) = false ];
+  repeated ContractCodeHistoryEntry contract_code_history = 4 [ (gogoproto.nullable) = false ];
```
Comment (alpe) [inline proto/cosmwasm/wasm/v1/genesis.proto:56]: 👍

File: x/wasm/keeper/keeper.go [hunk 4]
```diff
@@ -1074,11 +1074,11 @@ func (k Keeper) importContract(ctx sdk.Context, contractAddr sdk.AccAddress, c *
...
 	}
-	historyEntry := c.ResetFromGenesis(ctx)
-	k.appendToContractHistory(ctx, contractAddr, historyEntry)
+
+	k.appendToContractHistory(ctx, contractAddr, entries...)
 	k.storeContractInfo(ctx, contractAddr, c)
-	k.addToContractCodeSecondaryIndex(ctx, contractAddr, historyEntry)
-	k.addToContractCreatorSecondaryIndex(ctx, creatorAddress, historyEntry.Updated, contractAddr)
+	k.addToContractCodeSecondaryIndex(ctx, contractAddr, entries[len(entries)-1])
+	k.addToContractCreatorSecondaryIndex(ctx, creatorAddress, entries[0].Updated, contractAddr)
```
Comment (alpe) [inline x/wasm/keeper/keeper.go:1081]: Very nice to use the history data to feed the indexes with proper dates! 👍

File: x/wasm/types/ibc.pb.go [hunk 5]
```diff
@@ -5,21 +5,17 @@ package types
 
 import (
 	fmt "fmt"
+	_ "github.com/cosmos/gogoproto/gogoproto"
+	proto "github.com/gogo/protobuf/proto"
 	io "io"
 	math "math"
 	math_bits "math/bits"
-
```
Comment (alpe) [inline x/wasm/types/ibc.pb.go:11]: Please run `make proto-all` instead that includes the code formatters

File: x/wasm/keeper/genesis.go [hunk 6]
```diff
@@ -41,7 +41,7 @@ func InitGenesis(ctx sdk.Context, keeper *Keeper, data types.GenesisState, staki
 		if err != nil {
 			return nil, sdkerrors.Wrapf(err, "address in contract number %d", i)
 		}
-		err = keeper.importContract(ctx, contractAddr, &contract.ContractInfo, contract.ContractState)
+		err = keeper.importContract(ctx, contractAddr, &contract.ContractInfo, contract.ContractState, contract.ContractCodeHistory)
```
Comment (alpe) [inline x/wasm/keeper/genesis.go:44]: Please add some sanity checks for `ContractCodeHistory` to the ValidateBacis()

**Review by alpe — COMMENTED** [review]
This looks almost perfect now. Thanks a lot for your updates!
I have added some minor notes now but I need to do a deeper review tomorrow.

**Comment by GNaD13** [comment]
> This looks almost perfect now. Thanks a lot for your updates! I have added some minor notes now but I need to do a deeper review tomorrow.

Thank youuuuuuuuuuu sir :3

**Pull request merged**